From d62d43910608deecd447ec9cbbc7b21c01353f22 Mon Sep 17 00:00:00 2001
From: Archit Taneja <archit@ti.com>
Date: Wed, 25 Sep 2013 16:31:52 +0530
Subject: [PATCH 0544/1115] omapdss: hdmi pll: Add feature for DRA7x to enable
 PLL via external mux

OMAP5 and DRA7x HDMI PLLs defer in terms of how the registers can be accessed.
On OMAP5, the PLL registers  are always accessed via the DSS domain. On DRA7x,
the PLLs can also be accessed via an ocp2scp interface, in this case, we don't
need to enable the DSS clock domain.

There is a control module register which selects the mode of access. Add a
feature parameter which configures PLL access via the DSS clock domain.

Signed-off-by: Archit Taneja <archit@ti.com>
(cherry picked from commit 693a7db1a29fe8b768de5c04ae8be5a14ffd314c)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/video/omap2/dss/hdmi_pll.c |   27 ++++++++++++++++++++++++++-
 1 files changed, 26 insertions(+), 1 deletions(-)

diff --git a/drivers/video/omap2/dss/hdmi_pll.c b/drivers/video/omap2/dss/hdmi_pll.c
index ca8df27..21eb101 100644
--- a/drivers/video/omap2/dss/hdmi_pll.c
+++ b/drivers/video/omap2/dss/hdmi_pll.c
@@ -25,6 +25,8 @@ struct hdmi_pll_features {
 	bool sys_reset;
 	/* this is a hack, need to replace it with a better computation of M2 */
 	bool bound_dcofreq;
+	bool ext_mux_ctrl;
+
 	unsigned long fint_min, fint_max;
 	u16 regm_max;
 	unsigned long dcofreq_low_min, dcofreq_low_max;
@@ -184,6 +186,9 @@ int hdmi_pll_enable(struct hdmi_pll_data *pll, struct hdmi_wp_data *wp)
 {
 	u16 r = 0;
 
+	if (pll_feat->ext_mux_ctrl)
+		dss_dpll_enable_ctrl(DSS_DPLL_HDMI, true);
+
 	r = hdmi_wp_set_pll_pwr(wp, HDMI_PLLPWRCMD_ALLOFF);
 	if (r)
 		return r;
@@ -205,12 +210,16 @@ int hdmi_pll_enable(struct hdmi_pll_data *pll, struct hdmi_wp_data *wp)
 
 void hdmi_pll_disable(struct hdmi_pll_data *pll, struct hdmi_wp_data *wp)
 {
+	if (pll_feat->ext_mux_ctrl)
+		dss_dpll_enable_ctrl(DSS_DPLL_HDMI, false);
+
 	hdmi_wp_set_pll_pwr(wp, HDMI_PLLPWRCMD_ALLOFF);
 }
 
 static const struct hdmi_pll_features omap44xx_pll_feats = {
 	.sys_reset		=	false,
 	.bound_dcofreq		=	false,
+	.ext_mux_ctrl		=	false,
 	.fint_min		=	500000,
 	.fint_max		=	2500000,
 	.regm_max		=	4095,
@@ -223,6 +232,20 @@ static const struct hdmi_pll_features omap44xx_pll_feats = {
 static const struct hdmi_pll_features omap54xx_pll_feats = {
 	.sys_reset		=	true,
 	.bound_dcofreq		=	true,
+	.ext_mux_ctrl		=	false,
+	.fint_min		=	620000,
+	.fint_max		=	2500000,
+	.regm_max		=	2046,
+	.dcofreq_low_min	=	750000000,
+	.dcofreq_low_max	=	1500000000,
+	.dcofreq_high_min	=	1250000000,
+	.dcofreq_high_max	=	2500000000UL,
+};
+
+static const struct hdmi_pll_features dra7xx_pll_feats = {
+	.sys_reset		=	true,
+	.bound_dcofreq		=	true,
+	.ext_mux_ctrl		=	true,
 	.fint_min		=	620000,
 	.fint_max		=	2500000,
 	.regm_max		=	2046,
@@ -251,9 +274,11 @@ static int hdmi_pll_init_features(struct platform_device *pdev)
 		break;
 
 	case OMAPDSS_VER_OMAP5:
-	case OMAPDSS_VER_DRA7xx:
 		src = &omap54xx_pll_feats;
 		break;
+	case OMAPDSS_VER_DRA7xx:
+		src = &dra7xx_pll_feats;
+		break;
 
 	default:
 		return -ENODEV;
-- 
1.7.5.4

